### DISCUSSION TITLE
How to develop swiper plugin?

### DISCUSSION THREAD

**Discussion by linonetwo** [nolimits4web/swiper#4632] [Q&A] [answered]
I'm going to make a plugin similar to https://github.com/nolimits4web/swiper/blob/master/src/components/effect-coverflow/effect-coverflow.js and solve https://github.com/nolimits4web/Swiper/issues/1928 

But how to dev it locally and see changes? There is no an API to pass plugin to the swiper instance, so how to let it know we want it to load the effect-coverflow.js ?

**Comment by nolimits4web** [comment, marked as answer]
Setup any env you know, e.g. web pack, Vite, rollup, etc. Install swiper (`npm i swiper`), create a new plugin file (probably based on Coverflow.js)
```
import Swiper from 'swiper'
import myPlugin from './path/to/myplugin.js';

Swiper.use([myPlugin]);

new Swiper(...);
```

**Reply by linonetwo** [reply]
Sorry, I forget there is `Swiper.use([myPlugin]);`, I will make a try now.

**Reply by linonetwo** [reply]
BTW @nolimits4web , when dev in my project's environment , 

```
import $ from 'swiper/utils/dom';
import { bindModuleMethods } from 'swiper/utils/utils';
// or
import $ from 'swiper/cjs/utils/dom';
import { bindModuleMethods } from 'swiper/cjs/utils/utils';
```

produces `Cannot find module 'swiper/utils/dom' or its corresponding type declarations.ts(2307)`, do you have advice to solve this?

**Reply by nolimits4web** [reply]
you shouldn't rely and re-use Swiper internals

**Reply by linonetwo** [reply]
Okay, I see effect-coverflow use these internals.

I will try to use dom vanilla methods instead.

**Reply by linonetwo** [reply]
I can see basically internal elements are "Dom7 element", I believe they are created by `$` from `import $ from '../../utils/dom';` as how effect-coverflow.js is doing.

If I don't have access to the internal dom methods, how to do something like this?

```js
$shadowBeforeEl = $(
            `<div class="swiper-slide-shadow-${isHorizontal ? 'left' : 'top'}"></div>`,
          );
```

And I won't be able to do

```ts
         create() {
            const swiper = this;
            bindModuleMethods(swiper, {
              coverflowEffect: {
                ...Coverflow,
              },
            });
```

**Reply by linonetwo** [reply]
Can you expose dom utils and `bindModuleMethods` for plugin developers? If I'm going to PR them into your repo, I will refactor them back to relative import.

**Reply by nolimits4web** [reply]
> If I don't have access to the internal dom methods, how to do something like this?
```js
const el = document.createElement('div');
el.classList.add(`swiper-slide-shadow-${isHorizontal ? 'left' : 'top'}`)
```

**Reply by linonetwo** [reply]
OK, got it, they don't need to be "Dom7 element". And for `bindModuleMethods`, I think I can just copy the function code to my side. But it's better that swiper can export them somehow.

**Reply by nolimits4web** [reply]
That helper is also not needed, what it does is `swiper.myPlugin.someMethod = myPlugin.someMethod.bind(swiper)`
